tests/acme_endorsement.py

```diff
@@ -0,0 +1,341 @@
+                raise TimeoutError(
+                    f"Not all nodes had the correct ACME-endorsed TLS certificate installed after {timeout} seconds"
+                )
+
+            num_ok = 0
+            for node in network.nodes:
+                iface = node.host.rpc_interfaces["acme_endorsed_interface"]
+                try:
+                    context = ssl.SSLContext(ssl.PROTOCOL_TLSv1_2)
+                    context.verify_mode = ssl.CERT_REQUIRED
+                    context.check_hostname = True
+                    for crt in ca_certs:
+                        context.load_verify_locations(cadata=crt)
+
+                    s = socket()
+                    s.settimeout(1)
+                    c = context.wrap_socket(s, server_hostname=network_name)
+                    c.connect((iface.host, iface.port))
+                    cert_der = c.getpeercert(binary_form=True)
+                    cert = load_der_x509_certificate(cert_der, default_backend())
+                    if cert.subject.rfc4514_string() != "CN=" + network_name:
+                        raise Exception("Common name mismatch")
+                    cert_public_key = cert.public_key().public_bytes(
+                        encoding=Encoding.PEM, format=PublicFormat.SubjectPublicKeyInfo
+                    )
+                    if network_public_key != cert_public_key:
+                        raise Exception("Public key mismatch")
+                    num_ok += 1
+                except Exception as ex:
+                    LOG.trace(f"Likely expected exception: {ex}")
+
+            if num_ok != len(args.nodes):
+                time.sleep(1)
+
+        LOG.info(
+            f"Success: all nodes had correct certificates installed after {int(time.time() - start_time)} seconds"
+        )
+
+
+def get_binary(url, filename):
+    if not os.path.isfile(filename):
+        urllib.request.urlretrieve(
+            url,
+            filename,
+        )
+        os.chmod(filename, 0o744)
+
+
+def start_mock_dns(filename, listen_address, mgmt_address, out, err, env=None):
+    p = subprocess.Popen(
+        [
+            filename,
+            "-http01",
+            "",
+            "-https01",
+            "",
+            "-dns01",
+            listen_address,
+            "-tlsalpn01",
+            "",
+            "-management",
+            mgmt_address,
+        ],
+        stdout=out,
+        stderr=err,
+        close_fds=True,
+        env=env,
+    )
+    time.sleep(1)
+    return p
+
+
+def register_endorsed_hosts(args, network_name, dns_mgmt_address):
+    endorsed_hosts = [
+        node.rpc_interfaces["acme_endorsed_interface"].host for node in args.nodes
+    ]
+    data = str(
+        json.dumps(
+            {
+                "host": network_name,
+                "addresses": endorsed_hosts,
+            }
+        )
+    ).encode("utf-8")
+    urllib.request.urlopen("http://" + dns_mgmt_address + "/add-a", data=data)
```

Comment:
```suggestion
    urllib.request.urlopen(f"http://{dns_mgmt_address}/add-a", data=data)
```

This is really a matter of preference, but I find f-strings clearer and more compact? ymmv!